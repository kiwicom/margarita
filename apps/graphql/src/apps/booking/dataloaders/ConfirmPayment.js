// @flow

import stringify from 'json-stable-stringify';
import { OptimisticDataloader } from '@kiwicom/graphql-utils';

import fetch from '../../../services/fetch/tequilaFetch';
import type {
  ConfirmPaymentPayloadType,
  ConfirmPaymentOutputType,
  ConfirmPaymentAPIResponse,
} from '../Booking';

const parseParameters = (input: ConfirmPaymentPayloadType) => {
  return {
    transaction_id: input.transactionId,
    ...(input.bookingToken ? { booking_id: input.bookingToken } : {}),
  };
};

const parseResults = (
  results: $ReadOnlyArray<ConfirmPaymentAPIResponse>,
): $ReadOnlyArray<ConfirmPaymentOutputType> => {
  return results.map(result => ({
    isSuccess: !result.status,
    price: result.price,
  }));
};

const confirmPayment = async (
  parameters: $ReadOnlyArray<ConfirmPaymentPayloadType>,
) => {
  const method = 'POST';
  const results: $ReadOnlyArray<ConfirmPaymentAPIResponse> = await Promise.all(
    parameters.map(input => {
      return fetch('/v2/booking/confirm_payment', method, {
        body: JSON.stringify(parseParameters(input)),
      });
    }),
  );

  return parseResults(results);
};

export default function confirmPaymentLoader() {
  return new OptimisticDataloader(
    (
      keys: $ReadOnlyArray<ConfirmPaymentPayloadType>,
    ): Promise<$ReadOnlyArray<ConfirmPaymentOutputType | Error>> =>
      confirmPayment(keys),
    {
      cacheKeyFn: stringify,
    },
  );
}
