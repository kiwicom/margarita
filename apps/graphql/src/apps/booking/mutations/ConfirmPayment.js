// @flow

import type { GraphqlContextType } from '../../../services/graphqlContext/GraphQLContext';
import ConfirmPaymentPayload from '../types/inputs/ConfirmPaymentPayload';
import ConfirmPaymentOutput from '../types/outputs/ConfirmPayment';
import type { ConfirmPaymentPayloadType } from '../Booking';

type Args = {|
  +payload: ConfirmPaymentPayloadType,
|};

const ConfirmPayment = {
  name: 'ConfirmPayment',
  description: 'Confirms the payment for a given booking. (POST Request)',
  args: {
    payload: {
      type: ConfirmPaymentPayload,
    },
  },
  type: ConfirmPaymentOutput,
  resolve: async (_: mixed, args: Args, { dataLoader }: GraphqlContextType) => {
    const bookingConfirmation = await dataLoader.confirmPayment.load(
      args.payload,
    );
    return bookingConfirmation;
  },
};

export default ConfirmPayment;
