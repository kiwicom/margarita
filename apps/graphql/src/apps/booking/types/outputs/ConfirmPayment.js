// @flow

import { GraphQLObjectType, GraphQLBoolean, GraphQLFloat } from 'graphql';

const ConfirmPaymentOutput = new GraphQLObjectType({
  name: 'ConfirmPaymentOutput',
  fields: {
    isSuccess: {
      description: 'Indicates whether the payment was successful',
      type: GraphQLBoolean,
    },
    price: {
      description: 'The average price of the booking per passenger',
      type: GraphQLFloat,
    },
  },
});

export default ConfirmPaymentOutput;
