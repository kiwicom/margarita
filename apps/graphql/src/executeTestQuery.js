// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from './Schema';
import createContext from './services/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
  acceptLanguage: ?string,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
