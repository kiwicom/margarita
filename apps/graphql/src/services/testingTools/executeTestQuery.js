// @flow

import { graphql as originalGraphQL } from 'graphql';

import schema from '../../Schema';
import createContext from '../graphqlContext/GraphQLContext';

/**
 * Executes GraphQL query using our schema. This function should
 * be used only during testing because it uses faked context.
 */
export default function executeTestQuery(
  query: string,
  variables: ?Object,
): Promise<Object> {
  return originalGraphQL(schema, query, null, createContext(), variables);
}
