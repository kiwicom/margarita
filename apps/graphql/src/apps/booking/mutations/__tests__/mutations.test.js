// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import executeTestQuery from '../../../../services/testingTools/executeTestQuery';
import ConfirmPayment from './__datasets__/confirmPayment.json';

fetch.mockResponses([JSON.stringify(ConfirmPayment)]);

generateTestsFromFixtures(`${__dirname}/__fixtures__`, input => {
  return executeTestQuery(input);
});
