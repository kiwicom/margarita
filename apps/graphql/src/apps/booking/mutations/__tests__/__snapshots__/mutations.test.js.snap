// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: ConfirmPayment.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
mutation {
    confirmPayment(payload:{transactionId: "sandbox_5130968", bookingToken: "5130968"}){
        isSuccess
    }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "confirmPayment": {
      "isSuccess": true
    }
  }
}
`;
